File: bannerlayout/src/main/java/com/wkz/bannerlayout/widget/BannerTipsLayout.java
package com.wkz.bannerlayout.widget;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.v4.view.MarginLayoutParamsCompat;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.wkz.bannerlayout.annotation.TipsDotsSiteMode;
import com.wkz.bannerlayout.annotation.TipsLayoutSiteMode;
import com.wkz.bannerlayout.annotation.TipsProgressesSiteMode;
import com.wkz.bannerlayout.annotation.TipsTitleSiteMode;
import com.wkz.bannerlayout.utils.DisplayUtils;

public final class BannerTipsLayout extends RelativeLayout {

    private Context mContext;
    private TextView mTextView;
    private LinearLayout mDotsContainer;
    private LinearLayout mProgressesContainer;

    /**
     * Initialize the dots
     */
    public void setDots(@NonNull BannerTipsLayout.DotsInterface dotsInterface) {
        this.mDotsContainer.removeAllViews();

        for (int i = 0; i < dotsInterface.dotsCount(); ++i) {
            View view = new View(mContext);
            view.setBackground(dotsInterface.dotsSelector());
            view.setEnabled(i == 0);

            MarginLayoutParams marginLayoutParams = new MarginLayoutParams(
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsWidth()),
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsHeight()));
            MarginLayoutParamsCompat.setMarginStart(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsLeftMargin()));
            marginLayoutParams.topMargin = DisplayUtils.dp2px(mContext, dotsInterface.dotsTopMargin());
            MarginLayoutParamsCompat.setMarginEnd(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsRightMargin()));
            marginLayoutParams.bottomMargin = DisplayUtils.dp2px(mContext, dotsInterface.dotsBottomMargin());
            this.mDotsContainer.addView(view, marginLayoutParams);
        }

        this.mDotsContainer.measure(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);

        RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        for (int i = 0; i < dotsInterface.dotsSite().length; i++) {
            params.addRule(dotsInterface.dotsSite()[i]);
        }
        this.addView(this.mDotsContainer, params);
    }

    /**
     * Initialize the Progresses
     */
    public void setProgresses(@NonNull BannerTipsLayout.ProgressInterface progressInterface) {
        this.mProgressesContainer.removeAllViews();

        for (int i = 0; i < progressInterface.progressCount(); ++i) {
            View view = new View(mContext);
            view.setBackground(progressInterface.progressBuilder().build());

            MarginLayoutParams marginLayoutParams = new MarginLayoutParams(
                    (int) progressInterface.progressBuilder().getWidth(),
                    (int) progressInterface.progressBuilder().getHeight());
            MarginLayoutParamsCompat.setMarginStart(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, progressInterface.progressLeftMargin()));
            marginLayoutParams.topMargin = DisplayUtils.dp2px(mContext, progressInterface.progressTopMargin());
            MarginLayoutParamsCompat.setMarginEnd(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, progressInterface.progressRightMargin()));
            marginLayoutParams.bottomMargin = DisplayUtils.dp2px(mContext, progressInterface.progressBottomMargin());
            this.mProgressesContainer.addView(view, marginLayoutParams);
        }

        this.mProgressesContainer.measure(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);

        RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        for (int i = 0; i < progressInterface.progressSite().length; i++) {
            params.addRule(progressInterface.progressSite()[i]);
        }
        this.addView(this.mProgressesContainer, params);
    }

    @NonNull
    public FrameLayout.LayoutParams setBannerTips(@NonNull BannerTipsLayout.TipsInterface tipsInterface) {
        FrameLayout.LayoutParams tipsParams = new FrameLayout.LayoutParams(tipsInterface.tipsWidth(), tipsInterface.tipsHeight());
        @TipsLayoutSiteMode
        int tipsSiteMode = tipsInterface.tipsSite();
        switch (tipsSiteMode) {
            case TipsLayoutSiteMode.LEFT:
                tipsParams.gravity = Gravity.START;
                break;
            case TipsLayoutSiteMode.TOP:
                tipsParams.gravity = Gravity.TOP;
                break;
            case TipsLayoutSiteMode.RIGHT:
                tipsParams.gravity = Gravity.END;
                break;
            case TipsLayoutSiteMode.BOTTOM:
                tipsParams.gravity = Gravity.BOTTOM;
                break;
            case TipsLayoutSiteMode.CENTER:
                tipsParams.gravity = Gravity.CENTER;
        }

        if (tipsInterface.showBackgroundColor()) {
            this.setBackgroundColor(tipsInterface.tipsLayoutBackgroundColor());
        }

        return tipsParams;
    }

    /**
     * Update the dot position
     */
    public void changeDotsPosition(int position, int newPosition) {
        View childAt = this.mDotsContainer.getChildAt(position);
        View newChildAt = this.mDotsContainer.getChildAt(newPosition);
        if (childAt != null) {
            childAt.setEnabled(false);
        }

        if (newChildAt != null) {
            newChildAt.setEnabled(true);
        }

    }

    /**
     * Update the Progress position
     */
    public void changeProgressesPosition(int position, int newPosition) {
        View childAt = this.mProgressesContainer.getChildAt(position);
        View newChildAt = this.mProgressesContainer.getChildAt(newPosition);
        if (childAt != null) {
            if (childAt.getBackground() instanceof ProgressDrawable) {
                ((ProgressDrawable) childAt.getBackground()).end();
            }
        }

        if (newChildAt != null) {
            if (newChildAt.getBackground() instanceof ProgressDrawable) {
                ((ProgressDrawable) newChildAt.getBackground()).start();
            }
        }

    }

    /**
     * Update title, the default on the left
     */
    public final void setTitle(@NonNull BannerTipsLayout.TitleInterface titleInterface) {
        this.mTextView.setGravity(Gravity.CENTER_VERTICAL);
        this.mTextView.setTextColor(titleInterface.titleColor());
        this.mTextView.setTextSize(titleInterface.titleSize());
        this.mTextView.setSingleLine(true);
        this.mTextView.setEllipsize(TextUtils.TruncateAt.END);
        this.mTextView.setBackgroundColor(titleInterface.titleBackgroundColor());
        this.mTextView.setPadding(
                DisplayUtils.dp2px(mContext, titleInterface.titleLeftMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleTopMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleRightMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleBottomMargin())
        );

        RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(
                titleInterface.titleWidth() == -1 ?
                        ViewGroup.LayoutParams.MATCH_PARENT :
                        DisplayUtils.dp2px(mContext, titleInterface.titleWidth()),
                titleInterface.titleHeight() == -2 ?
                        ViewGroup.LayoutParams.WRAP_CONTENT :
                        DisplayUtils.dp2px(mContext, titleInterface.titleHeight())
        );
        for (int i = 0; i < titleInterface.titleSite().length; i++) {
            params.addRule(titleInterface.titleSite()[i]);
        }

        this.addView(this.mTextView, params);
    }

    public void setTitle(String title) {
        this.clearText();
        if (!TextUtils.isEmpty(title)) {
            this.mTextView.setText(title);
        }
    }

    private void clearText() {
        this.mTextView.setText("");
    }

    public BannerTipsLayout(@NonNull Context context) {
        super(context);
        this.mContext = context;
        this.mTextView = new TextView(mContext);
        this.mDotsContainer = new LinearLayout(mContext);
        this.mProgressesContainer = new LinearLayout(mContext);
    }

    public interface TipsInterface {
        boolean showBackgroundColor();

        int tipsSite();

        int tipsWidth();

        int tipsHeight();

        int tipsLayoutBackgroundColor();
    }


    public interface TitleInterface {
        int titleColor();

        float titleSize();

        float titleLeftMargin();

        float titleTopMargin();

        float titleRightMargin();

        float titleBottomMargin();

        float titleWidth();

        float titleHeight();

        @ColorInt
        int titleBackgroundColor();

        @TipsTitleSiteMode
        int[] titleSite();
    }

    public interface DotsInterface {
        int dotsCount();

        Drawable dotsSelector();

        float dotsHeight();

        float dotsWidth();

        float dotsLeftMargin();

        float dotsTopMargin();

        float dotsRightMargin();

        float dotsBottomMargin();

        @TipsDotsSiteMode
        int[] dotsSite();
    }

    public interface ProgressInterface {
        int progressCount();

        float progressLeftMargin();

        float progressTopMargin();

        float progressRightMargin();

        float progressBottomMargin();

        @TipsProgressesSiteMode
        int[] progressSite();

        ProgressDrawable.Builder progressBuilder();
    }
}

File: bannerlayout/src/main/java/com/fphoenixcorneae/bannerlayout/widget/BannerTipsLayout.kt
package com.fphoenixcorneae.bannerlayout.widget

import android.content.Context
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.core.view.MarginLayoutParamsCompat
import com.fphoenixcorneae.bannerlayout.annotation.TipsDotsSiteMode
import com.fphoenixcorneae.bannerlayout.annotation.TipsLayoutSiteMode
import com.fphoenixcorneae.bannerlayout.annotation.TipsProgressesSiteMode
import com.fphoenixcorneae.bannerlayout.annotation.TipsTitleSiteMode
import com.fphoenixcorneae.bannerlayout.utils.DisplayUtils

class BannerTipsLayout(private val mContext: Context) : RelativeLayout(mContext) {
    private val mTextView = TextView(mContext)
    private val mDotsContainer = FixedLinearLayout(mContext)
    private val mProgressesContainer = FixedLinearLayout(mContext)

    /**
     * Initialize the dots
     */
    fun setDots(dotsInterface: DotsInterface) {
        mDotsContainer.removeAllViews()
        for (i in 0 until dotsInterface.dotsCount()) {
            val view = View(mContext)
            view.background = dotsInterface.dotsSelector()
            view.isEnabled = i == 0
            val marginLayoutParams = MarginLayoutParams(
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsWidth()),
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsHeight()))
            MarginLayoutParamsCompat.setMarginStart(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsLeftMargin()))
            marginLayoutParams.topMargin = DisplayUtils.dp2px(mContext, dotsInterface.dotsTopMargin())
            MarginLayoutParamsCompat.setMarginEnd(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, dotsInterface.dotsRightMargin()))
            marginLayoutParams.bottomMargin = DisplayUtils.dp2px(mContext, dotsInterface.dotsBottomMargin())
            mDotsContainer.addView(view, marginLayoutParams)
        }
        mDotsContainer.measure(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        val params = LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        for (i in dotsInterface.dotsSite()!!.indices) {
            params.addRule(dotsInterface.dotsSite()!![i])
        }
        this.addView(mDotsContainer, params)
    }

    /**
     * Initialize the Progresses
     */
    fun setProgresses(progressInterface: ProgressInterface) {
        mProgressesContainer.removeAllViews()
        for (i in 0 until progressInterface.progressCount()) {
            val view = View(mContext)
            view.background = progressInterface.progressBuilder()!!.build()
            val marginLayoutParams = MarginLayoutParams(
                    progressInterface.progressBuilder()?.width?.toInt() ?: 0,
                    progressInterface.progressBuilder()?.height?.toInt() ?: 0)
            MarginLayoutParamsCompat.setMarginStart(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, progressInterface.progressLeftMargin()))
            marginLayoutParams.topMargin = DisplayUtils.dp2px(mContext, progressInterface.progressTopMargin())
            MarginLayoutParamsCompat.setMarginEnd(marginLayoutParams,
                    DisplayUtils.dp2px(mContext, progressInterface.progressRightMargin()))
            marginLayoutParams.bottomMargin = DisplayUtils.dp2px(mContext, progressInterface.progressBottomMargin())
            mProgressesContainer.addView(view, marginLayoutParams)
        }
        mProgressesContainer.measure(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        val params = LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        for (i in progressInterface.progressSite()!!.indices) {
            params.addRule(progressInterface.progressSite()!![i])
        }
        this.addView(mProgressesContainer, params)
    }

    fun setBannerTips(tipsInterface: TipsInterface): FrameLayout.LayoutParams {
        val tipsParams = FrameLayout.LayoutParams(tipsInterface.tipsWidth(), tipsInterface.tipsHeight())
        @TipsLayoutSiteMode val tipsSiteMode = tipsInterface.tipsSite()
        when (tipsSiteMode) {
            TipsLayoutSiteMode.LEFT -> tipsParams.gravity = Gravity.START
            TipsLayoutSiteMode.TOP -> tipsParams.gravity = Gravity.TOP
            TipsLayoutSiteMode.RIGHT -> tipsParams.gravity = Gravity.END
            TipsLayoutSiteMode.BOTTOM -> tipsParams.gravity = Gravity.BOTTOM
            TipsLayoutSiteMode.CENTER -> tipsParams.gravity = Gravity.CENTER
        }
        if (tipsInterface.showBackgroundColor()) {
            setBackgroundColor(tipsInterface.tipsLayoutBackgroundColor())
        }
        return tipsParams
    }

    /**
     * Update the dot position
     */
    fun changeDotsPosition(position: Int, newPosition: Int) {
        val childAt = mDotsContainer.getChildAt(position)
        val newChildAt = mDotsContainer.getChildAt(newPosition)
        if (childAt != null) {
            childAt.isEnabled = false
        }
        if (newChildAt != null) {
            newChildAt.isEnabled = true
        }
    }

    /**
     * Update the Progress position
     */
    fun changeProgressesPosition(position: Int, newPosition: Int) {
        val childAt = mProgressesContainer.getChildAt(position)
        val newChildAt = mProgressesContainer.getChildAt(newPosition)
        if (childAt != null) {
            if (childAt.background is ProgressDrawable) {
                (childAt.background as ProgressDrawable).end()
            }
        }
        if (newChildAt != null) {
            if (newChildAt.background is ProgressDrawable) {
                (newChildAt.background as ProgressDrawable).start()
            }
        }
    }

    /**
     * stop the Progress
     */
    fun stopProgresses(position: Int) {
        val childAt = mProgressesContainer.getChildAt(position)
        if (childAt != null) {
            if (childAt.background is ProgressDrawable) {
                (childAt.background as ProgressDrawable).stop()
            }
        }
    }

    /**
     * Update title, the default on the left
     */
    fun setTitle(titleInterface: TitleInterface) {
        mTextView.gravity = Gravity.CENTER_VERTICAL
        mTextView.setTextColor(titleInterface.titleColor())
        mTextView.textSize = titleInterface.titleSize()
        mTextView.isSingleLine = true
        mTextView.ellipsize = TextUtils.TruncateAt.END
        mTextView.setBackgroundColor(titleInterface.titleBackgroundColor())
        mTextView.setPadding(
                DisplayUtils.dp2px(mContext, titleInterface.titleLeftMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleTopMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleRightMargin()),
                DisplayUtils.dp2px(mContext, titleInterface.titleBottomMargin())
        )
        val params = LayoutParams(
                if (titleInterface.titleWidth() == -1f) ViewGroup.LayoutParams.MATCH_PARENT else DisplayUtils.dp2px(mContext, titleInterface.titleWidth()),
                if (titleInterface.titleHeight() == -2f) ViewGroup.LayoutParams.WRAP_CONTENT else DisplayUtils.dp2px(mContext, titleInterface.titleHeight())
        )
        for (i in titleInterface.titleSite()!!.indices) {
            params.addRule(titleInterface.titleSite()!![i])
        }
        this.addView(mTextView, params)
    }

    fun setTitle(title: String?) {
        clearText()
        if (!TextUtils.isEmpty(title)) {
            mTextView.text = title
        }
    }

    private fun clearText() {
        mTextView.text = ""
    }

    interface TipsInterface {
        fun showBackgroundColor(): Boolean
        fun tipsSite(): Int
        fun tipsWidth(): Int
        fun tipsHeight(): Int
        fun tipsLayoutBackgroundColor(): Int
    }

    interface TitleInterface {
        fun titleColor(): Int
        fun titleSize(): Float
        fun titleLeftMargin(): Float
        fun titleTopMargin(): Float
        fun titleRightMargin(): Float
        fun titleBottomMargin(): Float
        fun titleWidth(): Float
        fun titleHeight(): Float

        @ColorInt
        fun titleBackgroundColor(): Int

        @TipsTitleSiteMode
        fun titleSite(): IntArray?
    }

    interface DotsInterface {
        fun dotsCount(): Int
        fun dotsSelector(): Drawable
        fun dotsHeight(): Float
        fun dotsWidth(): Float
        fun dotsLeftMargin(): Float
        fun dotsTopMargin(): Float
        fun dotsRightMargin(): Float
        fun dotsBottomMargin(): Float

        @TipsDotsSiteMode
        fun dotsSite(): IntArray?
    }

    interface ProgressInterface {
        fun progressCount(): Int
        fun progressLeftMargin(): Float
        fun progressTopMargin(): Float
        fun progressRightMargin(): Float
        fun progressBottomMargin(): Float

        @TipsProgressesSiteMode
        fun progressSite(): IntArray?
        fun progressBuilder(): ProgressDrawable.Builder?
    }
}